pkg/reconciler/coordination/deployment/deployment_controller.go

```diff
@@ -0,0 +1,370 @@
+			return kubeClusterClient.AppsV1().Deployments().Cluster(clusterName).Namespace(namespace)
+		},
+	}
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), controllerName)
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			enqueue(obj, c.upstreamViewQueue, logger.WithValues("view", "upstream"))
+			enqueue(obj, c.syncerViewQueue, logger.WithValues("view", "syncer"))
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldObj, ok := old.(coordination.Object)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("resource should be a coordination.Object, but was %T", oldObj))
+				return
+			}
+			newObj, ok := new.(coordination.Object)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("resource should be a coordination.Object, but was %T", newObj))
+				return
+			}
+
+			if coordination.AnySyncerViewChanged(oldObj, newObj) {
+				enqueue(new, c.syncerViewQueue, logger.WithValues("view", "syncer"))
+			}
+			if coordination.UpstreamViewChanged(oldObj, newObj, deploymentContentsEqual) {
+				enqueue(new, c.upstreamViewQueue, logger.WithValues("view", "upstream"))
+			}
+		},
+	})
+
+	return c, nil
+}
+
+// controller reconciles watches deployments and coordinates them between SyncTargets
+type controller struct {
+	upstreamViewQueue workqueue.RateLimitingInterface
+	syncerViewQueue   workqueue.RateLimitingInterface
+
+	getDeployment func(clusterName logicalcluster.Name, namespace, name string) (*appsv1.Deployment, error)
+	patcher       func(clusterName logicalcluster.Name, namespace string) committer.Patcher[*appsv1.Deployment]
+
+	syncerViewRetriever coordination.SyncerViewRetriever[*appsv1.Deployment]
+	gvr                 schema.GroupVersionResource
+}
+
+func (c *controller) committer(clusterName logicalcluster.Name, namespace string) CommitFunc {
+	return committer.NewCommitterScoped[*Deployment, Patcher, *DeploymentSpec, *DeploymentStatus](c.patcher(clusterName, namespace))
+}
+
+func filter[K comparable, V interface{}](aMap map[K]V, keep func(key K) bool) map[K]V {
+	result := make(map[K]V)
+	for key, val := range aMap {
+		if keep(key) {
+			result[key] = val
+		}
+	}
+	return result
+}
+
+func deploymentContentsEqual(old, new interface{}) bool {
+	oldDeployment, ok := old.(*appsv1.Deployment)
+	if !ok {
+		return false
+	}
+	newDeployment, ok := new.(*appsv1.Deployment)
+	if !ok {
+		return false
+	}
+	oldAnnotations := filter(oldDeployment.Annotations, func(key string) bool {
+		return !strings.HasPrefix(key, v1alpha1.ClusterSpecDiffAnnotationPrefix)
+	})
+	newAnnotations := filter(newDeployment.Annotations, func(key string) bool {
+		return !strings.HasPrefix(key, v1alpha1.ClusterSpecDiffAnnotationPrefix)
+	})
+
+	return equality.Semantic.DeepEqual(oldDeployment.Labels, newDeployment.Labels) &&
+		equality.Semantic.DeepEqual(oldAnnotations, newAnnotations) &&
+		oldDeployment.Spec.Replicas == newDeployment.Spec.Replicas
```

Comment:
You'll need to safely deference the pointers and compare values

Reply:
Specifically, `Replicas`

Reply:
Oh Yes. Silly me !

Reply:
Fixed it, and added related unit tests in commit https://github.com/kcp-dev/kcp/pull/2336/commits/ce95ecf96dbf0ff1faccbab4d7f89129c8d71def